internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,324 @@
+		r.targetResolverState.Endpoints = []resolver.Endpoint{{Addresses: []resolver.Address{{Addr: target.Endpoint()}}}}
+		r.targetResolverReady = true
+	} else {
+		wcc := &wrappingClientConn{
+			stateListener: r.updateTargetResolverState,
+			parent:        r,
+		}
+		if r.targetResolver, err = targetResolverBuilder.Build(target, wcc, opts); err != nil {
+			return nil, fmt.Errorf("delegating_resolver: unable to build the resolver for target %s: %v", target, err)
+		}
+	}
+
+	if r.proxyResolver, err = r.proxyURIResolver(opts); err != nil {
+		return nil, fmt.Errorf("delegating_resolver: failed to build resolver for proxy URL %q: %v", r.proxyURL, err)
+	}
+	return r, nil
+}
+
+// proxyURIResolver creates a resolver for resolving proxy URIs using the
+// "dns" scheme. It adjusts the proxyURL to conform to the "dns:///" format and
+// builds a resolver with a wrappingClientConn to capture resolved addresses.
+func (r *delegatingResolver) proxyURIResolver(opts resolver.BuildOptions) (resolver.Resolver, error) {
+	proxyBuilder := resolver.Get("dns")
+	if proxyBuilder == nil {
+		panic("delegating_resolver: resolver for proxy not found for scheme dns")
+	}
+	r.proxyURL.Scheme = "dns"
+	r.proxyURL.Path = "/" + r.proxyURL.Host
+	r.proxyURL.Host = "" // Clear the Host field to conform to the "dns:///" format
+
+	proxyTarget := resolver.Target{URL: *r.proxyURL}
+	wcc := &wrappingClientConn{
+		stateListener: r.updateProxyResolverState,
+		parent:        r,
+	}
+	return proxyBuilder.Build(proxyTarget, wcc, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+		r.targetResolver = nil
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+		r.proxyResolver = nil
+	}
+}
+
+// combinedAddressesIfReadyLocked creates a list of combined addresses by
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) combinedAddressesIfReadyLocked() ([]resolver.Address, []resolver.Endpoint, bool) {
+	if !r.targetResolverReady || !r.proxyResolverReady {
+		return nil, nil, false
+	}
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetResolverState.Addresses {
+			newAddr := resolver.Address{Addr: proxyAddr.Addr}
+			var user url.Userinfo
+			if r.proxyURL.User != nil {
+				user = *r.proxyURL.User
+			}
+			newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+				User:        user,
+				ConnectAddr: targetAddr.Addr,
+			})
+			addresses = append(addresses, newAddr)
+		}
```

Comment:
Optional: This could be made a little more concise and yet be as readable:

```golang
		for _, targetAddr := range r.targetResolverState.Addresses {
			var user url.Userinfo
			if r.proxyURL.User != nil {
				user = *r.proxyURL.User
			}
			addresses = append(proxyAddr,  proxyattributes.Options{
				User:        user,
				ConnectAddr: targetAddr.Addr,
			})
		}
```